### ISSUE TITLE
Groovy Closures inaccessible

### ISSUE THREAD

**Issue by pickypg** [elastic/elasticsearch#16194]
This starts with ES 2.2 as we move to the script modules. The below script works in ES 2.1, but not ES 2.2 (the impact of which could be pretty big):

```http
PUT /test/type/1
{
  "message": "This is a message!",
  "numbers" : [1, 2, 3, 4]
}

PUT /test/type/2
{
  "message": "This is a message! too",
  "numbers" : [4, 2, 3, 5]
}

GET /test/_search
{
  "script_fields": {
    "use_closure": {
      "script": {
        "inline": "doc['numbers'].values.findAll { it % 2 == 0 }"
      }
    }
  }
}
```

I noticed that the issue seemed to be ClassNotFound errors related to the security manager, so I started adding permissions until I hit a wall because it could not access the generated closure's class.

```
  permission org.elasticsearch.script.ClassPermission "groovy.lang.Closure";
  permission org.elasticsearch.script.ClassPermission "org.codehaus.groovy.reflection.CachedMethod";
  permission org.elasticsearch.script.ClassPermission "org.codehaus.groovy.reflection.ClassInfo";
  permission org.elasticsearch.script.ClassPermission "org.codehaus.groovy.runtime.DefaultGroovyMethodsSupport";
  permission org.elasticsearch.script.ClassPermission "org.codehaus.groovy.runtime.GeneratedClosure";
```

cc @rasroh

**Comment by jasontedor** [comment]
Relates apache/groovy#248.

**Comment by jasontedor** [comment]
Closed via bdddea2dd0ecd3624e5484b3a34e668ca74cd6d0 from #16196.
